query/aql_nonaggr_batchexecutor.go

```diff
@@ -39,53 +40,57 @@ func (e *NonAggrBatchExecutorImpl) project() {
 	// wait for stream to clean up non used buffer before final aggregation
 	memutils.WaitForCudaStream(e.stream, e.qc.Device)
 	e.qc.OOPK.currentBatch.cleanupBeforeAggregation()
-
-	if e.qc.OOPK.currentBatch.resultSize >= e.qc.Query.Limit {
-		e.qc.OOPK.done = true
-	}
-}
-
-func (e *NonAggrBatchExecutorImpl) reduce() {
-	// nothing need to do for non-aggregation query
-}
-
-func (e *NonAggrBatchExecutorImpl) expandDimensions(numDims queryCom.DimCountsPerDimWidth) {
-	bc := &e.qc.OOPK.currentBatch
-	if bc.size == 0 {
-		//nothing to do
-		return
-	}
-
-	if bc.baseCountD.isNull() {
-		// baseCountD is null, no uncompression is needed
-		asyncCopyDimensionVector(bc.dimensionVectorD[1].getPointer(), bc.dimensionVectorD[0].getPointer(), bc.size, bc.resultSize,
-			numDims, bc.resultCapacity, bc.resultCapacity, memutils.AsyncCopyDeviceToDevice, e.stream, e.qc.Device)
-		bc.resultSize += bc.size
-		return
-	}
-
-	e.qc.doProfile(func() {
-		e.qc.OOPK.currentBatch.expand(numDims, bc.size, e.stream, e.qc.Device)
-		e.qc.reportTimingForCurrentBatch(e.stream, &e.start, expandEvalTiming)
-	}, "expand", e.stream)
 }
 
 func (e *NonAggrBatchExecutorImpl) prepareForDimEval(
 	dimRowBytes int, numDimsPerDimWidth queryCom.DimCountsPerDimWidth, stream unsafe.Pointer) {
 
 	bc := &e.qc.OOPK.currentBatch
+	lenWanted := e.qc.Query.Limit - e.qc.numberOfRowsWritten
+	if bc.size > lenWanted {
+		bc.size = lenWanted
+	}
+
 	if bc.resultCapacity == 0 {
-		bc.resultCapacity = e.qc.Query.Limit
+		bc.resultCapacity = bc.size
```

Comment:
in this case, if we don't have a limit for the query. resultCapacity should be the maximum size of all batches, not the first batch size

Reply:
I think the buffer capacity should be the size of the current batch after pre filtering. 
to reduce confusion, maybe I should not reuse the `resultCapacity` field.

Reply:
actually you are right, since we allocate buffer for current and next.

Reply:
+1